x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java

```diff
@@ -242,7 +242,7 @@ private boolean checkIndexAvailable(ClusterState state) {
             return false;
         }
         final IndexRoutingTable routingTable = state.routingTable().index(metadata.getIndex());
-        if (routingTable == null || routingTable.allPrimaryShardsActive() == false) {
+        if (routingTable == null || routingTable.readyForSearch(state) == false) {
             logger.debug("Index [{}] is not yet active", aliasName);
```

Comment:
This might be a semantic change. The method is used to populated the `isAvailable` value. Today is available means all primary shards are functional in stateful. It means available for both read and write. The proposed change means the index can be searchable (read) but not necessary writable because primary shards might not be active. There is a test case for it in this PR
> 2 unassigned primaries that are index only with some replicas that are all available

I wonder whether we should instead tighten this check to be both available and searchable instead of just searchable since there are [use cases](https://github.com/elastic/elasticsearch/blob/554b64e36ccc4859254fd26700dd68f532ecb61e/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java#L1882-L1884) where we use this check for write purpose. Or maybe we should have separate methods for available and searchable checks. But that may be a larger change. 

Another minor issue is that when we [report error](https://github.com/elastic/elasticsearch/blob/567ab3a29996dc31d3c8039a74309346f72669d8/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java#L138-L141) for `isAvailable == false`, it will still say "at least one primary shard is unavailable" even though it might not be the case.

Reply:
> The proposed change means the index can be searchable (read) but not necessary writable because primary shards might not be active.

In stateful: All replicas can be promoted to primary and are marked as DEFAULT so either the primary or replica will pass the `isSearchable()` check. The `isAvailable()` check is correct for writes except for the period of time after primary failure but before replica promotion. If there are no replicas to promote then the `isAvailable()` check will (correctly) return false. Since this check is more of an optimization / log helper it is does not need to be precisely correct during that short period.  Also, the setup for the test referenced is only applicable for stateless since it uses index/search only shards . 

In stateless: This is more relevant since as mentioned we use this check for both read and write and the shards are not all DEFAULT. However, read dominates the usage both by occurrence in the code and frequency of calls. The check is currently broken for reads which means that we are looking at primary shards to know if we can read. The consequence of that is reduced availability (for read requests when index shards are unavailable) and excessive logging that this PR fixes that. Fixing the read check comes at the cost of breaking the write check with the consequence is reduced availability (for write requests when search shards are unavailable ). Checking for both read/write would defeat the goal of this PR resulting in reduced availability for both read and write workflows. I think it is worth the effort split the check up into isAvailableForSearch, isAvailableForIndex, and isAvailableForRealTimeGet since in serverless each have their own semantics and can result in reduced availability if/when the index/search shards are unavailable. I was hoping to do that in a follow up PR since that is a large for this PR and this fixes the dominate issue. If you prefer I can pull the security changes here and handle it all once, but I do have a minor preference to proceed as-is with a follow up PR to handle the refactoring. 

>  it will still say "at least one primary shard is unavailable" even though it might not be the case.

Good catch ... will fix.

Reply:
> Checking for both read/write would defeat the goal of this PR resulting in reduced availability for both read and write workflows.

This prompted me to look into the context of this PR and I think it is ES-6842. IIUC, we have two excessive logs that this PR aims to fix:
1. (10%) "The security index is not yet available - no role mappings can be loaded "
2. (90%) "failed to load role mappings from index [.security] skipping all mappings."

For the first logging message, I think the fix is incorrect because the intention is that we don't want this log message when the security index is not "created". Therefore I think the fix should be adding a check for `indexExists()`. We call this method as a separate check in all other places (see this [example](https://github.com/elastic/elasticsearch/blob/ccc896d1284cb7a5fb7713110aedc1f8976f45c8/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java#L1642-L1643)) but it is missed for `getMappings()`. If the security index has not been created at all, it is not an exceptional situation and we always just return an empty result set. Concretely I am suggesting the following change to the `getMappings` method as the fix
```diff
     private void getMappings(ActionListener<List<ExpressionRoleMapping>> listener) {
-        if (securityIndex.isAvailable()) {
+        if (securityIndex.indexExists() == false) {
+            listener.onResponse(List.of());
+        } else if (securityIndex.isAvailable()) {
             loadMappings(listener);
         } else {

```

For the second logging message, I took a closer look at how primary and replica shards behave in edge cases. My conclusion is that the new `readyForSearch()` check actually behaves as what I suggested [above](https://github.com/elastic/elasticsearch/pull/99890#discussion_r1341268755), i.e. it checks _both_ available and searchable. When a primary shard fails, if no replica can be promoted to the new primary, [all replicas will be moved to "Unassigned"](https://github.com/elastic/elasticsearch/blob/ccc896d1284cb7a5fb7713110aedc1f8976f45c8/server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java#L659-L664). An unassigned shard will fail to pass the new `readyForSearch()` check which requires a shard to be active in the first place. Put it in another way, we cannot expect GET or Search to keep working when the security index has only search-only shards. Related, an index-only shard is *not* able to serve GET requests (realtime or not) when there is no search shards because the request fails earlier on at routing phase due to no search shards (whether this is really what we want in long term is a separate question). In summary, for GET or Search to reliably work, _both_ index and search shards are required to be available. There is no availability optimisation we can _currently_ have to serve either request with just an index shard or just a search shard.

Reply:
Thanks for digging in deep on this ... w.r.t. availability concerns I see you are correct, but still find it surprising that the availability of index_only shards impacts the availability of the search_only shards.  I suspect that is not what we want in the long term and will adjust (the security bits) of this PR accordingly.  I misinterpreted a comment that led me down this path and then applied some faulty assumptions. Thanks again for setting me straight ! (I'll take another look at the log message too.)

--- tangential ...
> the availability of index_only shards impacts the availability of the search_only shards

I don't think the inverse applies ... if the search_only shards are unavailable but the index_only shards can still be available. This (edge case) scenario could result in failure of read your writes via real time get where the write succeeds but the get fails. Not related here, but noteworthy.

Reply:
This change is still security related. Do you mean to keep it in for this PR?

Reply:
doh ! ... nope.. will revert that change too.